File: python/ivit/train/callbacks.py
"""
Training callbacks for monitoring and controlling the training process.
"""

from typing import Dict, Any, Optional, Callable
from pathlib import Path
from abc import ABC, abstractmethod
import numpy as np
import logging
import time
import json

logger = logging.getLogger(__name__)


class TrainingCallback(ABC):
    """
    Abstract base class for training callbacks.

    Callbacks can hook into various stages of training:
    - on_train_start: Called when training begins
    - on_train_end: Called when training ends
    - on_epoch_start: Called at the start of each epoch
    - on_epoch_end: Called at the end of each epoch
    - on_batch_start: Called before each batch
    - on_batch_end: Called after each batch
    """

    def on_train_start(self, trainer: 'Trainer', **kwargs) -> None:
        """Called when training starts."""
        pass

    def on_train_end(self, trainer: 'Trainer', **kwargs) -> None:
        """Called when training ends."""
        pass

    def on_epoch_start(self, trainer: 'Trainer', epoch: int, **kwargs) -> None:
        """Called at the start of each epoch."""
        pass

    def on_epoch_end(
        self,
        trainer: 'Trainer',
        epoch: int,
        metrics: Dict[str, float],
        **kwargs
    ) -> None:
        """Called at the end of each epoch."""
        pass

    def on_batch_start(
        self,
        trainer: 'Trainer',
        batch_idx: int,
        **kwargs
    ) -> None:
        """Called before each batch."""
        pass

    def on_batch_end(
        self,
        trainer: 'Trainer',
        batch_idx: int,
        loss: float,
        **kwargs
    ) -> None:
        """Called after each batch."""
        pass


class EarlyStopping(TrainingCallback):
    """
    Stop training when a monitored metric stops improving.

    Args:
        monitor: Metric to monitor (default: "val_loss")
        patience: Number of epochs with no improvement to wait
        min_delta: Minimum change to qualify as improvement
        mode: "min" or "max" (whether lower or higher is better)

    Examples:
        >>> early_stop = EarlyStopping(monitor="val_loss", patience=5)
        >>> trainer.fit(callbacks=[early_stop])
    """

    def __init__(
        self,
        monitor: str = "val_loss",
        patience: int = 5,
        min_delta: float = 0.0,
        mode: str = "min",
    ):
        self.monitor = monitor
        self.patience = patience
        self.min_delta = min_delta
        self.mode = mode

        self._best_value: Optional[float] = None
        self._counter = 0
        self._stopped_epoch = 0

    def on_train_start(self, trainer: 'Trainer', **kwargs) -> None:
        self._best_value = None
        self._counter = 0

    def on_epoch_end(
        self,
        trainer: 'Trainer',
        epoch: int,
        metrics: Dict[str, float],
        **kwargs
    ) -> None:
        current = metrics.get(self.monitor)
        if current is None:
            logger.warning(f"EarlyStopping: metric '{self.monitor}' not found")
            return

        if self._best_value is None:
            self._best_value = current
            return

        if self._is_improvement(current):
            self._best_value = current
            self._counter = 0
        else:
            self._counter += 1
            if self._counter >= self.patience:
                trainer._stop_training = True
                self._stopped_epoch = epoch
                logger.info(f"EarlyStopping: stopped at epoch {epoch}")

    def _is_improvement(self, current: float) -> bool:
        if self.mode == "min":
            return current < self._best_value - self.min_delta
        return current > self._best_value + self.min_delta

    def __repr__(self) -> str:
        return f"EarlyStopping(monitor='{self.monitor}', patience={self.patience})"


class ModelCheckpoint(TrainingCallback):
    """
    Save model checkpoints during training.

    Args:
        filepath: Path template for saving (can include {epoch}, {val_loss}, etc.)
        monitor: Metric to monitor for best model
        save_best_only: Only save when monitored metric improves
        mode: "min" or "max"

    Examples:
        >>> checkpoint = ModelCheckpoint(
        ...     filepath="checkpoints/model_{epoch:02d}.pt",
        ...     monitor="val_loss",
        ...     save_best_only=True,
        ... )
    """

    def __init__(
        self,
        filepath: str = "checkpoint.pt",
        monitor: str = "val_loss",
        save_best_only: bool = True,
        mode: str = "min",
    ):
        self.filepath = filepath
        self.monitor = monitor
        self.save_best_only = save_best_only
        self.mode = mode

        self._best_value: Optional[float] = None
        self._best_path: Optional[str] = None

    def on_train_start(self, trainer: 'Trainer', **kwargs) -> None:
        self._best_value = None
        # Create directory if needed
        Path(self.filepath).parent.mkdir(parents=True, exist_ok=True)

    def on_epoch_end(
        self,
        trainer: 'Trainer',
        epoch: int,
        metrics: Dict[str, float],
        **kwargs
    ) -> None:
        current = metrics.get(self.monitor)

        # Format filepath with metrics
        filepath = self.filepath.format(epoch=epoch, **metrics)

        if self.save_best_only:
            if current is None:
                logger.warning(f"ModelCheckpoint: metric '{self.monitor}' not found")
                return

            if self._best_value is None or self._is_improvement(current):
                self._best_value = current
                self._best_path = filepath
                self._save_checkpoint(trainer, filepath, epoch, metrics)
                logger.info(f"ModelCheckpoint: saved best model to {filepath}")
        else:
            self._save_checkpoint(trainer, filepath, epoch, metrics)
            logger.debug(f"ModelCheckpoint: saved to {filepath}")

    def _is_improvement(self, current: float) -> bool:
        if self.mode == "min":
            return current < self._best_value
        return current > self._best_value

    def _save_checkpoint(
        self,
        trainer: 'Trainer',
        filepath: str,
        epoch: int,
        metrics: Dict[str, float],
    ) -> None:
        """Save checkpoint to file."""
        try:
            import torch

            checkpoint = {
                'epoch': epoch,
                'model_state_dict': trainer.model.state_dict(),
                'optimizer_state_dict': trainer.optimizer.state_dict(),
                'metrics': metrics,
                'config': trainer.config,
            }
            torch.save(checkpoint, filepath)
        except ImportError:
            # Fallback: save metrics only
            with open(filepath + ".json", 'w') as f:
                json.dump({
                    'epoch': epoch,
                    'metrics': metrics,
                }, f, indent=2)

    @property
    def best_path(self) -> Optional[str]:
        """Get path to best saved model."""
        return self._best_path

    def __repr__(self) -> str:
        return f"ModelCheckpoint(filepath='{self.filepath}', monitor='{self.monitor}')"


class ProgressLogger(TrainingCallback):
    """
    Log training progress to console.

    Args:
        log_frequency: How often to log (every N batches)

    Examples:
        >>> logger = ProgressLogger(log_frequency=10)
    """

    def __init__(self, log_frequency: int = 10):
        self.log_frequency = log_frequency
        self._epoch_start_time: float = 0
        self._batch_losses: list = []

    def on_epoch_start(self, trainer: 'Trainer', epoch: int, **kwargs) -> None:
        self._epoch_start_time = time.time()
        self._batch_losses = []
        logger.info(f"Epoch {epoch + 1}/{trainer.epochs}")

    def on_batch_end(
        self,
        trainer: 'Trainer',
        batch_idx: int,
        loss: float,
        **kwargs
    ) -> None:
        self._batch_losses.append(loss)

        if (batch_idx + 1) % self.log_frequency == 0:
            avg_loss = np.mean(self._batch_losses[-self.log_frequency:])
            logger.info(f"  Batch {batch_idx + 1}: loss={avg_loss:.4f}")

    def on_epoch_end(
        self,
        trainer: 'Trainer',
        epoch: int,
        metrics: Dict[str, float],
        **kwargs
    ) -> None:
        elapsed = time.time() - self._epoch_start_time
        avg_loss = np.mean(self._batch_losses) if self._batch_losses else 0

        metrics_str = ", ".join([f"{k}={v:.4f}" for k, v in metrics.items()])
        logger.info(f"Epoch {epoch + 1} completed in {elapsed:.1f}s - loss={avg_loss:.4f}, {metrics_str}")

    def __repr__(self) -> str:
        return f"ProgressLogger(log_frequency={self.log_frequency})"


class LRScheduler(TrainingCallback):
    """
    Learning rate scheduler callback.

    Args:
        scheduler_type: Type of scheduler ("step", "cosine", "plateau")
        **kwargs: Arguments for the scheduler

    Examples:
        >>> # Step decay every 10 epochs
        >>> scheduler = LRScheduler("step", step_size=10, gamma=0.1)
        >>>
        >>> # Cosine annealing
        >>> scheduler = LRScheduler("cosine", T_max=100)
        >>>
        >>> # Reduce on plateau
        >>> scheduler = LRScheduler("plateau", patience=5, factor=0.5)
    """

    def __init__(self, scheduler_type: str = "step", **kwargs):
        self.scheduler_type = scheduler_type
        self.scheduler_kwargs = kwargs
        self._scheduler = None

    def on_train_start(self, trainer: 'Trainer', **kwargs) -> None:
        try:
            import torch.optim.lr_scheduler as lr_scheduler

            if self.scheduler_type == "step":
                self._scheduler = lr_scheduler.StepLR(
                    trainer.optimizer,
                    **self.scheduler_kwargs
                )
            elif self.scheduler_type == "cosine":
                self._scheduler = lr_scheduler.CosineAnnealingLR(
                    trainer.optimizer,
                    **self.scheduler_kwargs
                )
            elif self.scheduler_type == "plateau":
                self._scheduler = lr_scheduler.ReduceLROnPlateau(
                    trainer.optimizer,
                    **self.scheduler_kwargs
                )
            else:
                logger.warning(f"Unknown scheduler type: {self.scheduler_type}")
        except ImportError:
            logger.warning("PyTorch not available for LR scheduling")

    def on_epoch_end(
        self,
        trainer: 'Trainer',
        epoch: int,
        metrics: Dict[str, float],
        **kwargs
    ) -> None:
        if self._scheduler is None:
            return

        if self.scheduler_type == "plateau":
            val_loss = metrics.get('val_loss', metrics.get('loss', 0))
            self._scheduler.step(val_loss)
        else:
            self._scheduler.step()

        # Log current learning rate
        current_lr = trainer.optimizer.param_groups[0]['lr']
        logger.debug(f"Learning rate: {current_lr:.6f}")

    def __repr__(self) -> str:
        return f"LRScheduler(type='{self.scheduler_type}')"


class TensorBoardLogger(TrainingCallback):
    """
    Log training metrics to TensorBoard.

    Args:
        log_dir: Directory for TensorBoard logs

    Examples:
        >>> tb_logger = TensorBoardLogger(log_dir="runs/experiment1")
    """

    def __init__(self, log_dir: str = "runs"):
        self.log_dir = log_dir
        self._writer = None

    def on_train_start(self, trainer: 'Trainer', **kwargs) -> None:
        try:
            from torch.utils.tensorboard import SummaryWriter
            self._writer = SummaryWriter(log_dir=self.log_dir)
            logger.info(f"TensorBoard logging to {self.log_dir}")
        except ImportError:
            logger.warning("TensorBoard not available")

    def on_batch_end(
        self,
        trainer: 'Trainer',
        batch_idx: int,
        loss: float,
        **kwargs
    ) -> None:
        if self._writer is None:
            return

        global_step = trainer.current_epoch * trainer.batches_per_epoch + batch_idx
        self._writer.add_scalar('train/batch_loss', loss, global_step)

    def on_epoch_end(
        self,
        trainer: 'Trainer',
        epoch: int,
        metrics: Dict[str, float],
        **kwargs
    ) -> None:
        if self._writer is None:
            return

        for name, value in metrics.items():
            self._writer.add_scalar(f'metrics/{name}', value, epoch)

        # Log learning rate
        lr = trainer.optimizer.param_groups[0]['lr']
        self._writer.add_scalar('train/learning_rate', lr, epoch)

    def on_train_end(self, trainer: 'Trainer', **kwargs) -> None:
        if self._writer is not None:
            self._writer.close()

    def __repr__(self) -> str:
        return f"TensorBoardLogger(log_dir='{self.log_dir}')"

File: src/train/callbacks.cpp
/**
 * @file callbacks.cpp
 * @brief Training callbacks implementation
 */

#include "ivit/train/callbacks.hpp"
#include "ivit/train/trainer.hpp"

#include <iostream>
#include <fstream>
#include <iomanip>
#include <numeric>
#include <cmath>
#include <filesystem>

namespace ivit {
namespace train {

namespace fs = std::filesystem;

// ============================================================================
// EarlyStopping
// ============================================================================

EarlyStopping::EarlyStopping(
    const std::string& monitor,
    int patience,
    float min_delta,
    const std::string& mode
)
    : monitor_(monitor)
    , patience_(patience)
    , min_delta_(min_delta)
    , mode_(mode)
{}

void EarlyStopping::on_train_start(Trainer& trainer) {
    best_value_.reset();
    counter_ = 0;
    should_stop_ = false;
}

void EarlyStopping::on_epoch_end(
    Trainer& trainer,
    int epoch,
    const TrainingMetrics& metrics
) {
    float current = get_metric_value(metrics);
    if (std::isnan(current)) {
        std::cerr << "[EarlyStopping] Warning: metric '" << monitor_
                  << "' not found or is NaN" << std::endl;
        return;
    }

    if (!best_value_.has_value()) {
        best_value_ = current;
        return;
    }

    if (is_improvement(current)) {
        best_value_ = current;
        counter_ = 0;
    } else {
        counter_++;
        if (counter_ >= patience_) {
            trainer.stop_training();
            stopped_epoch_ = epoch;
            should_stop_ = true;
            std::cout << "[EarlyStopping] Stopped at epoch " << epoch
                      << " (best " << monitor_ << ": " << best_value_.value() << ")"
                      << std::endl;
        }
    }
}

bool EarlyStopping::is_improvement(float current) const {
    if (mode_ == "min") {
        return current < best_value_.value() - min_delta_;
    }
    return current > best_value_.value() + min_delta_;
}

float EarlyStopping::get_metric_value(const TrainingMetrics& metrics) const {
    if (monitor_ == "loss") return metrics.loss;
    if (monitor_ == "accuracy") return metrics.accuracy;
    if (monitor_ == "val_loss") return metrics.val_loss;
    if (monitor_ == "val_accuracy") return metrics.val_accuracy;
    return std::nanf("");
}

std::string EarlyStopping::repr() const {
    return "EarlyStopping(monitor='" + monitor_ + "', patience=" + std::to_string(patience_) + ")";
}

// ============================================================================
// ModelCheckpoint
// ============================================================================

ModelCheckpoint::ModelCheckpoint(
    const std::string& filepath,
    const std::string& monitor,
    bool save_best_only,
    const std::string& mode
)
    : filepath_(filepath)
    , monitor_(monitor)
    , save_best_only_(save_best_only)
    , mode_(mode)
{}

void ModelCheckpoint::on_train_start(Trainer& trainer) {
    best_value_.reset();
    best_path_.clear();

    // Create directory if needed
    fs::path dir = fs::path(filepath_).parent_path();
    if (!dir.empty() && !fs::exists(dir)) {
        fs::create_directories(dir);
    }
}

void ModelCheckpoint::on_epoch_end(
    Trainer& trainer,
    int epoch,
    const TrainingMetrics& metrics
) {
    float current = get_metric_value(metrics);
    std::string filepath = format_filepath(epoch, metrics);

    if (save_best_only_) {
        if (std::isnan(current)) {
            std::cerr << "[ModelCheckpoint] Warning: metric '" << monitor_
                      << "' not found" << std::endl;
            return;
        }

        if (!best_value_.has_value() || is_improvement(current)) {
            best_value_ = current;
            best_path_ = filepath;
            save_checkpoint(trainer, filepath, epoch, metrics);
            std::cout << "[ModelCheckpoint] Saved best model to " << filepath << std::endl;
        }
    } else {
        save_checkpoint(trainer, filepath, epoch, metrics);
    }
}

bool ModelCheckpoint::is_improvement(float current) const {
    if (mode_ == "min") {
        return current < best_value_.value();
    }
    return current > best_value_.value();
}

float ModelCheckpoint::get_metric_value(const TrainingMetrics& metrics) const {
    if (monitor_ == "loss") return metrics.loss;
    if (monitor_ == "accuracy") return metrics.accuracy;
    if (monitor_ == "val_loss") return metrics.val_loss;
    if (monitor_ == "val_accuracy") return metrics.val_accuracy;
    return std::nanf("");
}

std::string ModelCheckpoint::format_filepath(int epoch, const TrainingMetrics& metrics) const {
    std::string result = filepath_;

    // Replace {epoch} placeholder
    size_t pos = result.find("{epoch");
    if (pos != std::string::npos) {
        size_t end = result.find("}", pos);
        std::string format_spec = result.substr(pos, end - pos + 1);
        std::ostringstream oss;
        oss << std::setw(2) << std::setfill('0') << epoch;
        result.replace(pos, format_spec.length(), oss.str());
    }

    // Replace metric placeholders
    auto replace_metric = [&result](const std::string& name, float value) {
        size_t pos = result.find("{" + name);
        if (pos != std::string::npos) {
            size_t end = result.find("}", pos);
            std::string format_spec = result.substr(pos, end - pos + 1);
            std::ostringstream oss;
            oss << std::fixed << std::setprecision(4) << value;
            result.replace(pos, format_spec.length(), oss.str());
        }
    };

    replace_metric("loss", metrics.loss);
    replace_metric("accuracy", metrics.accuracy);
    replace_metric("val_loss", metrics.val_loss);
    replace_metric("val_accuracy", metrics.val_accuracy);

    return result;
}

void ModelCheckpoint::save_checkpoint(
    Trainer& trainer,
    const std::string& filepath,
    int epoch,
    const TrainingMetrics& metrics
) {
    trainer.save_checkpoint(filepath);
}

std::string ModelCheckpoint::repr() const {
    return "ModelCheckpoint(filepath='" + filepath_ + "', monitor='" + monitor_ + "')";
}

// ============================================================================
// ProgressLogger
// ============================================================================

ProgressLogger::ProgressLogger(int log_frequency)
    : log_frequency_(log_frequency)
{}

void ProgressLogger::on_epoch_start(Trainer& trainer, int epoch) {
    epoch_start_time_ = std::chrono::steady_clock::now();
    batch_losses_.clear();
    std::cout << "Epoch " << (epoch + 1) << "/" << trainer.epochs() << std::endl;
}

void ProgressLogger::on_batch_end(Trainer& trainer, int batch_idx, float loss) {
    batch_losses_.push_back(loss);

    if ((batch_idx + 1) % log_frequency_ == 0) {
        // Calculate average loss of recent batches
        size_t start = batch_losses_.size() > static_cast<size_t>(log_frequency_)
                       ? batch_losses_.size() - log_frequency_ : 0;
        float avg_loss = std::accumulate(
            batch_losses_.begin() + start, batch_losses_.end(), 0.0f
        ) / (batch_losses_.size() - start);

        std::cout << "  Batch " << (batch_idx + 1) << ": loss=" << std::fixed
                  << std::setprecision(4) << avg_loss << std::endl;
    }
}

void ProgressLogger::on_epoch_end(
    Trainer& trainer,
    int epoch,
    const TrainingMetrics& metrics
) {
    auto elapsed = std::chrono::steady_clock::now() - epoch_start_time_;
    double elapsed_sec = std::chrono::duration<double>(elapsed).count();

    float avg_loss = batch_losses_.empty() ? 0.0f :
        std::accumulate(batch_losses_.begin(), batch_losses_.end(), 0.0f) / batch_losses_.size();

    std::cout << "Epoch " << (epoch + 1) << " completed in "
              << std::fixed << std::setprecision(1) << elapsed_sec << "s"
              << " - loss=" << std::setprecision(4) << avg_loss
              << ", accuracy=" << std::setprecision(2) << metrics.accuracy << "%";

    if (metrics.val_loss > 0) {
        std::cout << ", val_loss=" << std::setprecision(4) << metrics.val_loss
                  << ", val_accuracy=" << std::setprecision(2) << metrics.val_accuracy << "%";
    }
    std::cout << std::endl;
}

std::string ProgressLogger::repr() const {
    return "ProgressLogger(log_frequency=" + std::to_string(log_frequency_) + ")";
}

// ============================================================================
// LRScheduler
// ============================================================================

LRScheduler::LRScheduler(
    const std::string& scheduler_type,
    int step_size,
    float gamma,
    int patience
)
    : scheduler_type_(scheduler_type)
    , step_size_(step_size)
    , gamma_(gamma)
    , patience_(patience)
{}

void LRScheduler::on_train_start(Trainer& trainer) {
    base_lr_ = trainer.config().learning_rate;
    current_lr_ = base_lr_;
    total_epochs_ = trainer.epochs();
    best_val_loss_.reset();
    plateau_counter_ = 0;
}

void LRScheduler::on_epoch_end(
    Trainer& trainer,
    int epoch,
    const TrainingMetrics& metrics
) {
    step(epoch, metrics.val_loss);
    trainer.set_learning_rate(current_lr_);
}

void LRScheduler::step(int epoch, float val_loss) {
    if (scheduler_type_ == "step") {
        update_lr_step(epoch);
    } else if (scheduler_type_ == "cosine") {
        update_lr_cosine(epoch);
    } else if (scheduler_type_ == "plateau") {
        update_lr_plateau(val_loss);
    } else if (scheduler_type_ == "exponential") {
        update_lr_exponential(epoch);
    }
}

void LRScheduler::update_lr_step(int epoch) {
    if ((epoch + 1) % step_size_ == 0) {
        current_lr_ *= gamma_;
    }
}

void LRScheduler::update_lr_cosine(int epoch) {
    // Cosine annealing
    current_lr_ = base_lr_ * 0.5f * (1.0f + std::cos(M_PI * epoch / total_epochs_));
}

void LRScheduler::update_lr_plateau(float val_loss) {
    if (std::isnan(val_loss)) return;

    if (!best_val_loss_.has_value() || val_loss < best_val_loss_.value()) {
        best_val_loss_ = val_loss;
        plateau_counter_ = 0;
    } else {
        plateau_counter_++;
        if (plateau_counter_ >= patience_) {
            current_lr_ *= gamma_;
            plateau_counter_ = 0;
            std::cout << "[LRScheduler] Reducing learning rate to " << current_lr_ << std::endl;
        }
    }
}

void LRScheduler::update_lr_exponential(int epoch) {
    current_lr_ = base_lr_ * std::pow(gamma_, epoch);
}

std::string LRScheduler::repr() const {
    return "LRScheduler(type='" + scheduler_type_ + "')";
}

// ============================================================================
// TensorBoardLogger
// ============================================================================

TensorBoardLogger::TensorBoardLogger(const std::string& log_dir)
    : log_dir_(log_dir)
{}

void TensorBoardLogger::on_train_start(Trainer& trainer) {
    // Create log directory
    if (!fs::exists(log_dir_)) {
        fs::create_directories(log_dir_);
    }
    initialized_ = true;
    std::cout << "[TensorBoardLogger] Logging to " << log_dir_ << std::endl;
}

void TensorBoardLogger::on_batch_end(Trainer& trainer, int batch_idx, float loss) {
    // Note: Full TensorBoard integration would require SummaryWriter
    // For now, this is a placeholder that could be extended
}

void TensorBoardLogger::on_epoch_end(
    Trainer& trainer,
    int epoch,
    const TrainingMetrics& metrics
) {
    // Log metrics to a simple text file that can be parsed
    std::ofstream file(log_dir_ + "/scalars.txt", std::ios::app);
    file << "epoch=" << epoch
         << " loss=" << metrics.loss
         << " accuracy=" << metrics.accuracy
         << " val_loss=" << metrics.val_loss
         << " val_accuracy=" << metrics.val_accuracy
         << " lr=" << metrics.learning_rate
         << std::endl;
}

void TensorBoardLogger::on_train_end(Trainer& trainer) {
    // Close writer if needed
}

std::string TensorBoardLogger::repr() const {
    return "TensorBoardLogger(log_dir='" + log_dir_ + "')";
}

// ============================================================================
// CSVLogger
// ============================================================================

CSVLogger::CSVLogger(const std::string& filepath, bool append)
    : filepath_(filepath), append_(append)
{}

void CSVLogger::on_train_start(Trainer& trainer) {
    // Create directory if needed
    fs::path dir = fs::path(filepath_).parent_path();
    if (!dir.empty() && !fs::exists(dir)) {
        fs::create_directories(dir);
    }

    // Write header if not appending or file doesn't exist
    if (!append_ || !fs::exists(filepath_)) {
        std::ofstream file(filepath_);
        file << "epoch,loss,accuracy,val_loss,val_accuracy,learning_rate,elapsed_seconds"
             << std::endl;
        header_written_ = true;
    }
}

void CSVLogger::on_epoch_end(
    Trainer& trainer,
    int epoch,
    const TrainingMetrics& metrics
) {
    std::ofstream file(filepath_, std::ios::app);
    file << epoch << ","
         << std::fixed << std::setprecision(6)
         << metrics.loss << ","
         << metrics.accuracy << ","
         << metrics.val_loss << ","
         << metrics.val_accuracy << ","
         << metrics.learning_rate << ","
         << metrics.elapsed_seconds
         << std::endl;
}

void CSVLogger::on_train_end(Trainer& trainer) {
    std::cout << "[CSVLogger] Training log saved to " << filepath_ << std::endl;
}

std::string CSVLogger::repr() const {
    return "CSVLogger(filepath='" + filepath_ + "')";
}

// ============================================================================
// LambdaCallback
// ============================================================================

LambdaCallback::LambdaCallback(
    EpochEndFn on_epoch_end_fn,
    BatchEndFn on_batch_end_fn
)
    : on_epoch_end_fn_(std::move(on_epoch_end_fn))
    , on_batch_end_fn_(std::move(on_batch_end_fn))
{}

void LambdaCallback::on_epoch_end(
    Trainer& trainer,
    int epoch,
    const TrainingMetrics& metrics
) {
    if (on_epoch_end_fn_) {
        on_epoch_end_fn_(trainer, epoch, metrics);
    }
}

void LambdaCallback::on_batch_end(Trainer& trainer, int batch_idx, float loss) {
    if (on_batch_end_fn_) {
        on_batch_end_fn_(trainer, batch_idx, loss);
    }
}

std::string LambdaCallback::repr() const {
    return "LambdaCallback()";
}

} // namespace train
} // namespace ivit
